test/python/circuit/test_controlled_gate.py

```diff
@@ -901,5 +901,20 @@ def test_deprecated_gates(self):
                 self.assertTrue(isinstance(old(*params), new))
 
 
+@ddt
+class TestControlledGateIssues(QiskitTestCase):
+    """Did you find a specific failing case? add it here with a reference to the issue."""
+    @data((RXGate, CRXGate),
+          (RYGate, CRYGate),
+          (RZGate, CRZGate))
+    @unpack
+    def test_rotation_with_ctrl_state(self, gate, controlled_gate):
+        """Test rotation with ctrl_state
+        See https://github.com/Qiskit/qiskit-terra/pull/4025
+        """
+        theta = 0.5
+        self.assertEqual(gate(theta).control(1, ctrl_state='1'), controlled_gate(theta))
```

Comment:
The equality check in ControlledGate may not be strong enough for this test. For instance it doesn't check that ctrl_state matches (bug) and doesn't check that the definitions match.

Reply:
Indeed. I think the problem of equivalence is beyond this PR (`assertEqual` for controlled gates are used all around these tests currently) and that unveiled other issues (like https://github.com/Qiskit/qiskit-terra/issues/4032). So I submitting this need as an issue in https://github.com/Qiskit/qiskit-terra/issues/4031

Reply:
There are some assertEquals on gates but more common is using `_compute_control_matrix` and comparing to Operator().data. This verifies that the definition produces the expected control matrix. 

Also, I'm not sure whether it is needed to have a narrowly defined, issue specific test like this when a more general test could be added (e.g. checking different ctrl_state for all standard gates)?

Reply:
`_compute_control_matrix` cannot be used in this case because https://github.com/Qiskit/qiskit-terra/issues/3304.

Yes, more cases can be added for all the gates. And that might find more bugs. This test is specifically for the problem that this PR is solving (the test do not pass in master). I can add tests for all the gates in a different PR (or submit issues for the cases that fail). Again, that's a different PR.

Reply:
Yeah as a matter of test organization, rather than adding ad-hoc cases for specific issues, can we think about what tests make sense and add them? This issue is pointing to a bug in functionality, so that functionality lacked an appropriate test. The tests should be tied to that functionality and generalized.

Reply:
> ...rather than adding ad-hoc cases for specific issues...

IMHO, having tests for ad-hoc issues is fine. They can help with track that you are not reintroducing a bug.

Anyway. 37c1ca266a83ea5e157ea10f884fecfd29d30ac9 extends the test to other gates and adds a note about changing the assertion onces #3304 is fixed.